### ISSUE TITLE
[javac] some differences ecj <-> javac regarding 'recent' features

### ISSUE THREAD

**Issue by stephan-herrmann** [eclipse-jdt/eclipse.jdt.core#2959]
While [taking stock](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2212#issuecomment-2323293285) of `run.javac` test results regarding Java 23 I observed the following differences which might be within reach for resolution:

* PrimitiveInPatternsTestSH
   * 1 javac bug see https://mail.openjdk.org/pipermail/compiler-dev/2024-September/027729.html
* MarkdownCommentTest:
   *  1 error, javac not complaining about unresolvable `@see` reference
* ModuleCompilationTests
    * 4 errors
* PatternMatching16Test -- is this a structural problem?
    * 83 failures, obviously not suitable for running back-to-back with javac23
    * 67 errors vs javac16
* RecordPatternTest 3 errors, javac-generated code throws ArithmeticException instead of expected MatchException (see comments in tests):
     * testGH1977_instance_field()
     * testGH1977_static_field()
     * testIssue1985_2()
* SwitchExpressionsYieldTest
   * currently 175 failures, many of which due to ill-configured javac invocation
   * remaining: 17 failures
* SwitchPatternTest: 5 errors (some related to wrongly configured --enable-preview, others not yet analysed)
* SwitchPatternTest21: 2 errors
      * testInternalDomination_this() (ecj no error, javac 2 errors)
      * testNaming() // javac jdk21 allows components to be named, but they can't be referenced

**Comment by stephan-herrmann** [comment]
> SwitchPatternTest21: 2 errors
> * testInternalDomination_this() (ecj no error, javac 2 errors)
> * testNaming() // javac jdk21 allows components to be named, but they can't be referenced

The class has been renamed to SwitchPatternTest22.

@srikanth-sankaran  are those differences of interest for your current work?

**testInternalDomination_this()**
* javac reports errors, that are not (yet?) detected by ecj:
```
X.java:5: error: this case label is dominated by a preceding case label
      case Object _, Integer _, X _ when o != null : System.out.println("Integer");
                     ^
X.java:5: error: this case label is dominated by a preceding case label
      case Object _, Integer _, X _ when o != null : System.out.println("Integer");
                                ^
2 errors
```
**testNaming()**:
* javac does not report the expected error, but only a warning:
```
X.java:6: warning: [fallthrough] possible fall-through into case
      default: System.out.println("Object");
      ^
1 warning
```
Test case has this comment _"javac jdk21 allows components to be named, but they can't be referenced"_.

When those are clarified, we can accept this test class into the paradise of successful run.javac comparison :)

**Comment by stephan-herrmann** [comment]
After #3218 also SwitchPatternTest is on the home stretch with just three test cases where javac reports errors which ecj doesn't see:
* testIssue3009_6()
   *  #3009
* testIssue3009_7() dto
* testIssue3031_2() 
   *  #3031
   * I dropped the ball from [this comment](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3031#issuecomment-2384755842)

And then the errors from #3009 look similar:
```
X.java:26: error: illegal start of type
		case W<Short>.NG.I<Integer>.S<Long, Integer, X> e -> 42;
		                                                ^
1 error
```

**Comment by stephan-herrmann** [comment]
After first successful builds I have set https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-23/ to run weekly (on weekends)

Currently these test classes have opted in:
* ImplicitlyDeclaredClassesTest
* ModuleImportTests
* PrimitivePatternTest
* PrimitivePatternTestSH
* SealedTypesTest
* SuperAfterStatementsTest

More classes will be admitted once they are green in run.javac mode.

**Comment by srikanth-sankaran** [comment]
> More classes will be admitted once they are green in run.javac mode.

Thanks a lot for doing this!

In my first innings as well, I never seem to have paid close attention (scratch that, any attention) to this capability. In fact I seem to have invented a completely new mechanism that is a lot less sophisticated, see `org.eclipse.jdt.core.tests.compiler.parser.AbstractSyntaxTreeTest.CHECK_JAVAC_PARSER` 

This is useful and I can participate in contributing and maintaining this in 2025 - after my reviews of Java 10-23 are over.

**Comment by stephan-herrmann** [comment]
Latest build https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-24/6/testReport/ (yes, focus is on **24** by now) shows a few failures in javac-comparison. This is with  "build 24-ea+22-2649".  Running the same tests locally with "build 24-ea+29-3578" those failures no longer occur. 

I filed https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/5463 which should eventually grant us a green build.

**Comment by stephan-herrmann** [comment]
We have a [green build](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-24/9/) with run.javac=OptIn and these test classes opting in:
* ImplicitlyDeclaredClassesTest
* InterfaceMethodsTest
* ModuleImportTests
* PrimitiveInPatternsTest
* PrimitiveInPatternsTestSH
* SealedTypesTests
* SuperAfterStatementsTest

**Comment by stephan-herrmann** [comment]
Recently, these tests had some issues 
* build failure due to releng issues and merge problems
* changes in javac behavior between 23 and 24 causing our comparison to complain
* ModuleImportTests had not been run (missing inclusion in TestAll).

All these have been resolved as of now, we have a green build again at https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-24/15/

**Comment by stephan-herrmann** [comment]
New test participating since https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3616
* PreviewFlagTest

**Comment by srikanth-sankaran** [comment]
> [@srikanth-sankaran](https://github.com/srikanth-sankaran) are those differences of interest for your current work?

Sorry for the late response. 

I believe `org.eclipse.jdt.core.tests.compiler.regression.SwitchPatternTest22.testNaming()` is a bug in javac and ECJ behavior is correct.

`org.eclipse.jdt.core.tests.compiler.regression.SwitchPatternTest22.testInternalDomination_this()` OTOH looks like an issue with us. I have opened a separate ticket : https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3626

**Comment by stephan-herrmann** [comment]
> > [@srikanth-sankaran](https://github.com/srikanth-sankaran) are those differences of interest for your current work?
> 
> Sorry for the late response.
> 
> I believe `org.eclipse.jdt.core.tests.compiler.regression.SwitchPatternTest22.testNaming()` is a bug in javac and ECJ behavior is correct.

I agree, 14.11.1 has this normative sentence:
> It is a compile-time error for a case label to have more than one case pattern and declare any pattern variables (other than those declared by a guard associated with the case label).

and this non-normative explanation:
> If a case label with more than one case pattern could declare pattern variables, then it would not be clear which variables would be initialized if the case label were to apply.

**Comment by stephan-herrmann** [comment]
For now I disabled https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-23/ after it showed a flood of failures, many (all?) of which are connected to #3586. Those failures got fixed in BETA_JAVA24 only (71b983181ac9f981340cd0fdfdf1def5fa391239).

Focus is on Java 24 anyway.

**Comment by stephan-herrmann** [comment]
After next merge from master, SwitchPatternTest22 wants to opt-in to run.javac mode, see #3626.

**Comment by stephan-herrmann** [comment]
While working on #3687 I see LocalEnumTest as another candidate to opt in to run.javac. With pending fixes I see exactly two differences vis a vis javac:
* test022: ecj complains that `case MX.BLEU` must be replaced with `BLEU`, javac is quiet
* test140: cast from enum type to `Runnable`: javac complains, ecj is quiet

@srikanth-sankaran @mpalat @jarthana  let me know if you recall any "recent" changes in either area. Thanks.

**Comment by srikanth-sankaran** [comment]
> * 
> 
> [@srikanth-sankaran](https://github.com/srikanth-sankaran) [@mpalat](https://github.com/mpalat) [@jarthana](https://github.com/jarthana) let me know if you recall any "recent" changes in either area. Thanks.

Please see 

https://bugs.openjdk.org/browse/JDK-8309372
https://bugs.openjdk.org/browse/JDK-8300542

Perhaps our implementation has gaps. Could you please raise a ticket after confirming this to be the case and assign it to me? I'll follow up.

BTW, I see that `org.eclipse.jdt.core.tests.compiler.regression.EnumTest.test022()` is not run for >= JDK21. This disabling happened on behalf of https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1252 - which is interesting. It looks like we will allow qualified enums while switching on sealed interfaces but now while switching on enumeration themselves.

**Comment by srikanth-sankaran** [comment]
See emission of `cannotUseQualifiedEnumConstantInCaseLabel()` in `org.eclipse.jdt.internal.compiler.ast.CaseStatement.resolveConstantLabel(BlockScope, TypeBinding, TypeBinding, Expression)`

**Comment by stephan-herrmann** [comment]
Not concerning a recent feature, but a hint I'd like to capture (from https://mail.openjdk.org/pipermail/compiler-dev/2025-February/029342.html):
> Note that capture conversion is always problematic when it comes to cast expressions -- because the cast has to "relax" some of the captured variable that might be introduced by capture conversion. This is an area of the specification that is underspecified, so javac (and other implementations) have to workaround such issues the best they can.

**Comment by stephan-herrmann** [comment]
InnerEmulationTest could opt in to run.javac once we fix #4081

**Comment by stephan-herrmann** [comment]
Meanwhile JDK 25 ea has graduated previous previews, which caused a few test failures on our side, but as of https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-25/9/testReport/ we're back on track.

**Comment by stephan-herrmann** [comment]
> InnerEmulationTest could opt in to run.javac once we fix [#4081](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/4081)

Done and https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-25/15 is green again.

_(I just had to include o.e.j.annotation in the build invocation, because it has a minor version bump since previous build)._

**Comment by stephan-herrmann** [comment]
Joining soon:
* DubiousOutcomeTest via https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4347

**Comment by stephan-herrmann** [comment]
Here's a fresh list of test classes currently participating:
* AssertionTest
* DubiousOutcomeTest
* ImplicitelyDeclaredClassesTest
* InnerEmulationTest
* InterfaceMethodsTest
* JavaNonLanguageTest
* LocalEnumTest
* ModuleImportTest
* PreviewFlagTest
* PrimitiveInPatternsTest
* PrimitiveInPatternsTestSH
* SealedTypesTests
* SuperAfterStatementsTest

This gives a total of 1347 tests successfully running with javac-comparison on Java 25.

**Comment by stephan-herrmann** [comment]
To simplify search for relevant issues in the OpenJDK bug tracker I started a new page as https://github.com/eclipse-jdt/eclipse.jdt.core/wiki/ECJ-Relevant-JDK-Issues

**Comment by stephan-herrmann** [comment]
Newly included in run.javac mode (opt-in):
* AnnotationTest (passes against javac 25 with jdt master)

New test job testing BETA_JAVA26 against javac 26 ea 
* https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-26/
  * failures in PreviewFlagTest will hopefully be fixed with #4831
  * for PrimitiveInPatternsTestSH we seem to need a newer JDK 26 EA, see https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/6723#note_6285208

**Comment by stephan-herrmann** [comment]
I shifted Java versions in run.javac tests:
* master at 26:
   * https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-26
   * currently one failure: new ecj-only warning since #5138, I added an excuse but accidentally did this in BETA_JAVA_27 :-/
* BETA_JAVA_27 at 27 ea:
   * https://ci.eclipse.org/jdt/job/eclipse.jdt.core-run.javac-27/
   * PASS

At this point we have a new member in run.javac mode:
* GenericsRegressionTest_9

**Comment by stephan-herrmann** [comment]
I'm about to add printing of statistics to the run.javac test mode.

Running locally, I get this output with `-Drun.javac=optIn`

at `-Dcompliance=26` and against javac 26:
```
Num tests with run.javac enabled: 1869
Num tests triggering a javac bug: 4
	JavacBug8016207
	JavacBug8387487
	JavacBugIvarInterning
	JavacHasWarningsEclipseNotConfigured
Num tests triggering an ecj bug : 1
	EclipseBug235809
```

at `-Dcompliance=24` and against javac 24:
```
Num tests with run.javac enabled: 1096
Num tests triggering a javac bug: 6
	JavacBug8016207
	JavacBug8361641
	JavacBug8365676
	JavacBug8387487
	JavacBugIvarInterning
	JavacHasWarningsEclipseNotConfigured
Num tests triggering an ecj bug : 0
```


at `-Dcompliance=21` and against javac 21:
```
Num tests with run.javac enabled: 1096
Num tests triggering a javac bug: 8
	JavacBug8016207
	JavacBug8337980
	JavacBug8343306
	JavacBug8361641
	JavacBug8365676
	JavacBug8387487
	JavacBugIvarInterning
	JavacHasWarningsEclipseNotConfigured
Num tests triggering an ecj bug : 0
```
* at 21 we have one instance of `Eclipse found error(s) but Javac did not find any` - an ecj bug at this compliance

at `-Dcompliance=11` and against javac 11:
```
Num tests with run.javac enabled: 615
Num tests triggering a javac bug: 6
	JavacBug8016207
	JavacBug8337980
	JavacBug8361641
	JavacBug8365676
	JavacBug8387487
	JavacBugIvarInterning
Num tests triggering an ecj bug : 0
```
* at 11 we also have two instances of `Javac found error(s) but Eclipse did not find any.` - assumed to be javac bugs.

at `-Dcompliance=1.8` and against javac 1.8:
```
Num tests with run.javac enabled: 507
Num tests triggering a javac bug: 1
	JavacBug8337980
Num tests triggering an ecj bug : 0
```
* at 1.8 we also have two instances of `Javac found error(s) but Eclipse did not find any.` - assumed to be javac bugs.


The difference between compliance levels essentially documents javac bugs coming and going over time. It seems that our book keeping of each bug's pivot version is more or less correct.

**Comment by stephan-herrmann** [comment]
In https://github.com/eclipse-jdt/eclipse.jdt.core/pull/5376  I changed two things for the benefit of this (long running) task:
* To let another test class opt in to run.javac mode all it takes now is adding `@RunJavac` to the class header
* A new test class `TestAllRunJavac` runs only those test classes having the `@RunJavac` annotation.
